content/blogs/gui-setup-part3-first-testing-problems.md

```diff
@@ -0,0 +1,267 @@
+
+Because of this failure, attempting to install such files will fail.
+This failure, coupled to the fact that the [custom setupapi file-copy queue callback function](https://learn.microsoft.com/en-us/windows/win32/setupapi/creating-a-custom-queue-callback-routine) I wrote in the GUI installer, did not correctly handle for file-copy failures (and instead was just requesting setupapi to re-attempt copying the same file), led to an infinite loop and thus an apparent hanging.
+
+#### The fix
+
+Fixing this bug is actually pretty simple.
+
+First, the [custom setupapi file-copy queue callback function](https://learn.microsoft.com/en-us/windows/win32/api/setupapi/nc-setupapi-psp_file_callback_w) in the GUI installer needs to recognize it is called due to a copy error
+(with the [`SPFILENOTIFY_COPYERROR`](https://learn.microsoft.com/en-us/windows/win32/setupapi/spfilenotify-copyerror) setupapi [notification](https://learn.microsoft.com/en-us/windows/win32/setupapi/notifications))
+and return `FILEOP_SKIP` or any other "error" value, instead of the default `FILEOP_DOIT` or `FILEOP_RETRY`.
+This would make the code more "future-proof" for any other sort of copy errors. (An improvement would also to show a custom error message box.)
+
+Then, in the `do_file_copyW()` function, instead of calling `GetTempPathW()` to get the temporary directory, just generate a temporary file path and name based on the target path, where the file will actually be installed.
+Since this path is guaranteed to exist on the system -- it is inside the write-accessible ReactOS installation partition and the (sub-)directory has been created by a previous step -- , a temporary file can be created there without problems.
+
+
+### 2nd main "bug"
+
+Now that the 1st bug has been fixed, retesting the GUI installer goes without apparent problems, until one notices the extreme slowness when copying files.
+Clearly this does not happen when using the text-mode USETUP.
+What is going on here?
+
+{{< video src="/img/blogs/hbelusca_gui-setup/guisetup_filecopy_slow.mp4" type="video/mp4" caption="File copy deadly slow">}}
+
+The GUI setup has been made to use the Wine-synced setupapi routines for extracting the files from the `reactos.cab` source.
+
+The gist of how the file copy loop works is as follows:
+
+1. Get the next file source/target paths from the queue;
+2. Call `do_file_copyW()` for the file: if it exists in the source installation directory, copy it and Goto Step 1.
+3. If not, check installation CAB file (`reactos.cab` in our case) with `extract_cabinet_file()`.
+If failed, throw error via the file-copy callback.
+4. Open CAB file, loop through it to find the file, extract it, and close the CAB file.
+5. Goto Step 1 for next file.
+
+In Steps 3. and 4., the helper function `extract_cabinet_file()` is used, and ultimately calls into the `cabinet.dll!FDICopy()` function (`cabinet.dll` is also a Wine-synced dll).
```

Comment:
```suggestion
In Steps 3 and 4, the helper function `extract_cabinet_file()` is used, and ultimately calls into the `cabinet.dll!FDICopy()` function (`cabinet.dll` is also a Wine-synced dll).
```
(I think it will be more readable this way)